File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/geofeature/Feature.java
/**
 * Copyright (c) Codice Foundation
 * <p>
 * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
 * General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or any later version.
 * <p>
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
 * is distributed along with this program and can be found at
 * <http://www.gnu.org/licenses/lgpl.html>.
 */
package org.codice.ddf.catalog.ui.query.geofeature;

import java.util.HashMap;
import java.util.Map;

/**
 * Base class for results returned from the FeatureService.
 */
public abstract class Feature {
    private String name;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public abstract String getType();

    /**
     * Subclasses should override with behavior specific to their geometry type
     * @return  A data object representing the Feature's geometry
     */
    public abstract Map<String, Object> getGeometryJsonObject();

    /**
     * Provides a JSON-serializable representation of the Feature
     * @return  A data object representing the Feature
     */
    public Map<String, Object> getJsonObject() {
        Map<String, Object> result = new HashMap<>();
        result.putAll(this.getGeometryJsonObject());
        result.put("name", this.name);
        result.put("type", this.getType());
        return result;
    }
}

File: platform/admin/ui/src/main/webapp/js/models/features/feature.js
/**
* Copyright (c) Codice Foundation
*
* This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
* General Public License as published by the Free Software Foundation, either version 3 of the
* License, or any later version.
    *
* This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
* even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
* Lesser General Public License for more details. A copy of the GNU Lesser General Public License
* is distributed along with this program and can be found at
* <http://www.gnu.org/licenses/lgpl.html>.
*
**/
/*global define*/
define(['backbone','jquery','underscore'], function (Backbone,$,_) {

    var Feature = {};

    var featureUrl = '/admin/jolokia/read/org.codice.ddf.admin.application.service.ApplicationService:service=application-service/AllFeatures';
    var featureByAppUrl = '/admin/jolokia/exec/org.codice.ddf.admin.application.service.ApplicationService:service=application-service/findApplicationFeatures/';
    var installUrl = '/admin/jolokia/exec/org.apache.karaf:type=feature,name=root/installFeature(java.lang.String,boolean)/';
    var uninstallUrl = '/admin/jolokia/exec/org.apache.karaf:type=feature,name=root/uninstallFeature(java.lang.String,boolean)/';



    Feature.Model = Backbone.Model.extend({
        initialize: function(options) {
            _.bindAll.apply(_, [this].concat(_.functions(this)));
            this.name = options.name;
        },
        urlRoot: function(){
            return featureUrl;
        },
        install: function(){
            return $.ajax({
                type: 'GET',
                url: installUrl + this.name + "/true",
                dataType: 'JSON'
            });
        },
        uninstall: function(){
            return $.ajax({
                type: 'GET',
                url: uninstallUrl + this.name + "/true",
                dataType: 'JSON'
            });
        }

    });

    Feature.Collection = Backbone.Collection.extend({
        model: Feature.Model,
        comparator: function(item) {
          return item.get('name');
        },
        initialize: function(options){
            this.type = options.type;
            this.appName = options.appName;
        },
        url: function() {
            if(this.type !== 'all'){
                return featureByAppUrl + this.appName;
            }else{
                return featureUrl;
            }

        },
        parse: function(resp){
            return resp.value;
        }
    });

    return Feature;
});

